From 0fff87982b7164a442b549509fa9fa792007880a Mon Sep 17 00:00:00 2001
From: Arjen Hiemstra <ahiemstra@heimr.nl>
Date: Tue, 15 Nov 2022 14:49:03 +0100
Subject: [PATCH] shell: Use the basic scene graph rendering loop on wayland

This avoids crashing Plasma when a surface gets destroyed too early
while still in use by the threaded loop. To avoid leaking things into
child processes, we clear the environment variable again after we've
created the initial views for the shell.

BUG: 447717


(cherry picked from commit 9c998d3083f622a1677782248d4c8e238c935dc2)
---
 shell/main.cpp        | 20 ++++++++++++++++++++
 shell/shellcorona.cpp |  2 ++
 shell/shellcorona.h   |  1 +
 3 files changed, 23 insertions(+)

diff --git a/shell/main.cpp b/shell/main.cpp
index b990c10d0f..f0db2a72b0 100644
--- a/shell/main.cpp
+++ b/shell/main.cpp
@@ -93,6 +93,16 @@ int main(int argc, char *argv[])
 
     bool replace = false;
 
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    if (app.platformName() == QStringLiteral("wayland")) {
+        // Use the basic rendering loop on wayland, because the threaded render
+        // loop can cause problems because a surface gets destroyed while still
+        // being used by the render thread.
+        // TODO KF6: Remove this as the problem has been fixed in Qt6
+        qputenv("QSG_RENDER_LOOP", "basic");
+    }
+#endif
+
     ShellCorona *corona;
     {
         QCommandLineParser cliOptions;
@@ -230,5 +240,15 @@ int main(int argc, char *argv[])
     corona->init();
     SoftwareRendererNotifier::notifyIfRelevant();
 
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    if (app.platformName() == QStringLiteral("wayland")) {
+        QObject::connect(corona, &ShellCorona::loaded, corona, []() {
+            // Clear the forced render loop so we avoid leaking it into other
+            // processes that get started by Plasma.
+            qunsetenv("QSG_RENDER_LOOP");
+        });
+    }
+#endif
+
     return app.exec();
 }
diff --git a/shell/shellcorona.cpp b/shell/shellcorona.cpp
index fa3ca442c0..ed470b349c 100644
--- a/shell/shellcorona.cpp
+++ b/shell/shellcorona.cpp
@@ -740,6 +740,8 @@ void ShellCorona::load()
         KConfigGroup coronaConfig(config(), "General");
         setImmutability((Plasma::Types::ImmutabilityType)coronaConfig.readEntry("immutability", static_cast<int>(Plasma::Types::Mutable)));
     }
+
+    Q_EMIT loaded();
 }
 
 void ShellCorona::primaryScreenChanged(QScreen *oldPrimary, QScreen *newPrimary)
diff --git a/shell/shellcorona.h b/shell/shellcorona.h
index 74d237e4a9..d4678e9974 100644
--- a/shell/shellcorona.h
+++ b/shell/shellcorona.h
@@ -141,6 +141,7 @@ Q_SIGNALS:
     void containmentPreviewReady(Plasma::Containment *containment, const QString &path);
     void accentColorFromWallpaperEnabledChanged();
     void colorChanged(const QColor &color);
+    void loaded();
 
 public Q_SLOTS:
     /**
-- 
GitLab

